ecl/hql/hqlopt.cpp

```diff
@@ -2195,19 +2259,27 @@ IHqlExpression * CTreeOptimizer::doCreateTransformed(IHqlExpression * transforme
     IHqlExpression * child = transformed->queryChild(0);
 
     //Any optimizations that remove the current node, or modify the current node don't need to check if the children are shared
-    //Removing child nodes could be included, but it may create more spillers/spliters - which may be significant in thor.
+    //Removing child nodes could be included, but it may create more spillers/splitters - which may be significant in thor.
     switch (op)
     {
     case no_if:
         {
             OwnedHqlExpr ret = optimizeIf(transformed);
+
+            //This won't split shared nodes, but one of the children may be shared - so processed here
+            if (!ret && transformed->isDataset())
+            {
+                //Convert IF(a, f(ds), g(ds)) to ds(IF(a,f,g)) - revisit since may increase dependencies
```

Comment:
This optimization is given precedence, yet it is not mentioned in either the jira issue or the PR summary.